src/services/organizationService.ts

```diff
@@ -490,6 +495,104 @@ export async function registerOrganization(
   }
 }
 
+export function addDelegate(
+  ipaCode: string,
+  userEmail: string
+): TaskEither<
+  // tslint:disable-next-line:max-union-size
+  IResponseErrorInternal | IResponseErrorNotFound | IResponseErrorConflict,
+  IResponseSuccessRedirectToResource<Organization, Organization>
+> {
+  const genericErrorHandler = (error: unknown) => {
+    log.error("An error occurred adding the delegate. %s", error);
+    return ResponseErrorInternal("An error occurred adding the delegate.");
+  };
+  return tryCatch<
+    IResponseErrorInternal | IResponseErrorNotFound | IResponseErrorConflict,
+    OrganizationModel
+  >(
+    () =>
+      OrganizationModel.findOne({
+        where: { ipaCode }
+      }),
+    genericErrorHandler
+  )
+    .chain(
+      fromPredicate<IResponseErrorNotFound, OrganizationModel>(
+        _ => _ !== null,
+        () =>
+          ResponseErrorNotFound(
+            "Not found",
+            "Registered organization does not exists"
+          )
+      )
+    )
+    .chain(
+      fromPredicate<
+        IResponseErrorConflict | IResponseErrorInternal,
+        OrganizationModel
+      >(
+        _ =>
+          _.registrationStatus ===
+          OrganizationRegistrationStatusEnum.REGISTERED,
+        () =>
+          ResponseErrorConflict(
+            "The organization has not completed the registration process yet."
+          )
+      )
+    )
+    .chain(organizationModel =>
+      tryCatch<IResponseErrorInternal, OrganizationModel>(
+        () =>
+          OrganizationUserModel.create({
+            createdAt: Date.now(),
+            organizationIpaCode: organizationModel.ipaCode,
+            updatedAt: Date.now(),
+            userEmail,
+            userRole: UserRoleEnum.ORG_DELEGATE
+          }).then(_ => organizationModel),
+        genericErrorHandler
+      )
+    )
+    .chain(organizationModel =>
+      tryCatch(
+        () =>
+          organizationModel.reload({
+            include: [
+              {
+                as: "users",
+                model: User
+              },
+              {
+                as: "legalRepresentative",
+                model: User
+              }
+            ],
+            order: [["users", "createdAt", "DESC"]]
+          }),
+        genericErrorHandler
+      )
+    )
+    .chain(organizationModel =>
+      fromEither(
+        toOrganizationObject(organizationModel).map(organization => {
+          const selfLink = organization.links.find(_ => _.rel === "self");
+          return ResponseSuccessRedirectToResource(
+            organization,
+            selfLink ? selfLink.href : "",
+            organization
+          );
+        })
+      ).mapLeft(errors => {
```

Comment:
brilliant ! I like this style of chaining tasks 👍 

(compare this against controller#addDelegate)